File: base/hourai/bot/extensions/validation/rejectors.py
import humanize
import re
from unidecode import unidecode
from datetime import datetime
from hourai import utils
from hourai.db import models
from .common import Validator, generalize_filter, split_camel_case


LOOSE_DELETED_USERNAME_MATCH = re.compile(r'(?i).*Deleted.*User.*')
TRANSFORMS = (lambda x: x, unidecode)


class NameMatchRejector(Validator):
    """A suspicion level validator that rejects users for username proximity to
    other users already on the server.
    """
    __slots__ = ("filter", "prefix", "subfield", "member_selector",
                 "min_match_length")

    def __init__(self, *, prefix, filter_func,
                 min_match_length=None, subfield=None, member_selector=None):
        self.filter = filter_func
        self.prefix = prefix
        self.subfield = subfield or (lambda m: m.name)
        self.member_selector = member_selector or (lambda m: m.name)
        self.min_match_length = min_match_length

    async def validate_member(self, ctx):
        member_names = {}
        for guild_member in filter(self.filter, ctx.guild.members):
            name = self.member_selector(guild_member) or ''
            member_names.update({
                p: generalize_filter(p) for p in self._split_name(name)
            })
        field_value = self.subfield(ctx.member)
        for filter_name, regex in member_names.items():
            if re.search(regex, field_value):
                ctx.add_rejection_reason(
                    self.prefix + f'Matches: `{filter_name}`')

    def _split_name(self, name):
        split_name = split_camel_case(name)
        if self.min_match_length is not None:
            split_name = (n for n in split_name
                          if len(n) >= self.min_match_length)
        return split_name


class StringFilterRejector(Validator):
    """A general validator that rejects users that have a field that matches
    a set of predefined list of regexes.
    """
    __slots__ = ("filters", "prefix", "match_func", "subfield", "use_transforms")

    def __init__(self, *, prefix, filters, full_match=False, subfield=None,
                 use_transforms=True):
        self.prefix = prefix or ''
        self.filters = [(f, re.compile(generalize_filter(f))) for f in filters]
        self.use_transforms = use_transforms
        self.subfield = subfield or (
            lambda ctx: (u.name for u in ctx.usernames))
        if full_match:
            self.match_func = lambda r: r.match
        else:
            self.match_func = lambda r: r.search

    async def validate_member(self, ctx):
        transforms = TRANSFORMS if self.use_transforms else (lambda x: x,)
        for field_value in self.subfield(ctx):
            for filter_name, regex in self.filters:
                for transform in transforms:
                    transformed = transform(field_value)
                    if self.match_func(regex)(transformed):
                        ctx.add_rejection_reason(
                            self.prefix + f'Matches: `{filter_name}`')


class NewAccountRejector(Validator):
    """A suspicion level validator that rejects users that were recently
    created.
    """
    __slots__ = ("lookback")

    def __init__(self, *, lookback):
        self.lookback = lookback

    async def validate_member(self, ctx):
        if ctx.member.created_at > datetime.utcnow() - self.lookback:
            lookback_naturalized = humanize.naturaltime(self.lookback)
            ctx.add_rejection_reason(
                f"Account created less than {lookback_naturalized}")


class DeletedAccountRejector(Validator):
    """A suspicion level validator that rejects users that are deleted or have
    tell-tale warning signs of faking a deleted account in the past.
    """
    __slots__ = ()

    async def validate_member(self, ctx):
        if utils.is_deleted_user(ctx.member):
            ctx.add_rejection_reason(
                "Deleted users cannot be active on Discord. User has been "
                "deleted by Discord of their own accord or for Trust and "
                "Safety reasons, or is faking account deletion.")

        for transform in TRANSFORMS:
            for username in ctx.usernames:
                name = transform(username.name)
                is_deleted = utils.is_deleted_username(username.name)
                if LOOSE_DELETED_USERNAME_MATCH.match(username.name) and \
                   not is_deleted:
                    ctx.add_rejection_reason(
                        f'"{username.name}" does not match Discord\'s deletion '
                        f'patterns. User may have attempted to fake account '
                        f"deletion.")
                elif is_deleted and username.discriminator is not None and \
                        username.discriminator < 100:
                    ctx.add_rejection_reason(
                        f'"{username.name}#{username.discriminator}" has an '
                        f'unusual discriminator. These are randomly generated. '
                        f'User may have attempted to fake account deletion.')


class NoAvatarRejector(Validator):
    """A suspicion level validator that rejects users without avatars."""
    __slots__ = ()

    async def validate_member(self, ctx):
        if ctx.member.avatar is None:
            ctx.add_rejection_reason("User has no avatar.")


class BannedUserRejector(Validator):
    """A malice level validator that rejects users that are banned on other
    servers.
    """
    __slots__ = ("min_guild_size")

    def __init__(self, *, min_guild_size):
        self.min_guild_size = min_guild_size

    async def validate_member(self, ctx):
        bans = await ctx.bot.storage.bans.get_user_bans(ctx.member.id)
        valid_bans = list(filter(self._is_valid_ban, bans))
        if len(valid_bans) <= 0:
            return
        reasons = set(b.reason for b in valid_bans if b.HasField('reason'))
        reason = f"Banned from {len(bans)} servers " if len(bans) > 1 else \
                 "Banned from another server "
        if len(reasons) > 0:
            reason += "for the following reasons:\n"
            reason += "\n    - ".join(reasons)
        ctx.add_rejection_reason(reason)

    def _is_valid_ban(self, ban):
        return not ban.guild_blocked and ban.guild_size >= self.min_guild_size


class BannedUsernameRejector(Validator):
    """A malice level validator that rejects users that share characteristics
    with banned users on the server:
     - Exact username matches (ignoring repeated whitespace and casing).
     - Exact avatar matches.
    """
    __slots__ = ()

    async def validate_member(self, ctx):
        if not ctx.guild.me.guild_permissions.ban_members:
            return
        bans = await ctx.bot.storage.bans.get_guild_bans(ctx.guild.id)
        self.__check_usernames(ctx, bans)
        self.__check_avatars(ctx, bans)

    def __check_avatars(self, ctx, bans):
        avatar = ctx.member.avatar
        if avatar is None:
            return
        for ban in bans:
            if not ban.HasField('avatar') or avatar != ban.avatar:
                continue
            reason = f"Exact username match with banned user: `{str(ban.user)}`."
            if ban.HasField('reason'):
                reason += f" Ban Reason: {ban.reason}"
            ctx.add_rejection_reason(reason)

    def __check_usernames(self, ctx, bans):
        ban_ids = [ban.user_id for ban in bans]
        ban_reasons = {ban.user_id:
                       ban.reason if ban.HasField('reason') else None
                       for ban in bans}
        with ctx.bot.create_storage_session() as session:
            matches = session.query(models.Username) \
                             .filter(models.Username.user_id.in_(ban_ids)) \
                             .distinct(models.Username.name)

            for transform in TRANSFORMS:
                normalized_usernames = set(self._normalize(transform(u.name))
                                           for u in ctx.usernames)
                # Don't match on empty string matches
                normalized_usernames.discard("")

                for banned_username in matches.all():
                    transformed = transform(banned_username.name)
                    normalized = self._normalize(transformed)
                    if not normalized in normalized_usernames:
                        continue
                    ban_reason = ban_reasons.get(banned_username.user_id)
                    reason = f"Exact avatar match with banned user: " + \
                             f"{banned_username.name}`."
                    if ban_reason is not None:
                        reason += f" Ban Reason: {ban_reason}"
                    ctx.add_rejection_reason(reason)
                    break

    def _normalize(self, val):
        return " ".join(val.casefold().split())


class LockdownRejector(Validator):
    """A malice level validator that rejects all users if the guild has been put
    into lockdown.
    """
    __slots__ = ()

    async def validate_member(self, ctx):
        guild_state = ctx.bot.guild_states[ctx.guild.id]
        if guild_state.is_locked_down:
            ctx.add_rejection_reason(
                'Lockdown enabled. All new joins must be manually verified.')

File: logger/src/validation/rejectors.rs
use async_trait::async_trait;
use chrono::Duration;
use chrono::offset::Utc;
use crate::cache::InMemoryCache;
use crate::db::{SqlPool, Username, Ban, ValidationBan};
use crate::models::Snowflake;
use dashmap::DashMap;
use std::collections::HashMap;
use regex::Regex;
use super::{*, context};
use twilight_model::user::User;

lazy_static! {
    static ref DELETED_USERNAME_MATCH: Regex = Regex::new("Deleted User [0-9a-fA-F]{8}").unwrap();
    static ref LOOSE_DELETED_USERNAME_MATCH: Regex = Regex::new("(?i).*Deleted.*User.*").unwrap();
}

fn is_user_deleted(user: &User) -> bool {
    user.avatar.is_none() && DELETED_USERNAME_MATCH.is_match(user.name.as_str())
}

struct DeletedUserRejector(SqlPool);

#[async_trait]
impl Validator for DeletedUserRejector {
    async fn validate(&self, ctx: &mut context::ValidationContext) -> Result<()> {
        if is_user_deleted(&ctx.member().user) {
            ctx.add_rejection_reason(
                "Deleted users cannot be active on Discord. User has been \
                deleted by Discord of their own accord or for Trust and \
                Safety reasons, or is faking account deletion.");
        }

        let usernames: Vec<Username> = Username::fetch(ctx.member().user.id, None)
                                                .fetch_all(&self.0)
                                                .await?;

        for username in usernames {
            let name = username.name.as_str();
            let is_deleted = DELETED_USERNAME_MATCH.is_match(name);
            if !is_deleted && LOOSE_DELETED_USERNAME_MATCH.is_match(name) {
                ctx.add_rejection_reason(
                    format!("\"{}\" does not match Discord\'s deletion patterns. User may have \
                             attemtped to fake account deletion.", name));
            } else if is_deleted && username.discriminator.map(|d| d < 100).unwrap_or(false) {
                ctx.add_rejection_reason(
                    format!("\"{}#{:04}\" has a unusual discriminator for a deleted user. These \
                             are randomly generated. User may have attemtped to fake account \
                             deletion.", name, username.discriminator.unwrap()));
            }
        }

        Ok(())
    }
}

struct BannedUserRejector {
    sql: SqlPool,
    cache: InMemoryCache,
    min_guild_size: u64
}

#[async_trait]
impl Validator for BannedUserRejector {
    async fn validate(&self, ctx: &mut context::ValidationContext) -> Result<()> {
        let bans: Vec<Ban> = Ban::fetch_user_bans(ctx.member().user.id)
                                 .fetch_all(&self.sql)
                                 .await?;

        let mut reasons: Vec<Option<String>> = Vec::new();
        for ban in bans {
            if let Some(guild) = self.cache.guild(ban.guild_id()) {
                if guild.member_count.unwrap_or(0) < self.min_guild_size {
                    continue;
                }
            } else {
                continue;
            }
            reasons.push(ban.reason);
        }

        if reasons.len() > 0 {
            let mut reason = format!("Banned from {} servers", reasons.len());
            let list: String =
                reasons.into_iter()
                       .filter_map(|r| r)
                       .collect::<Vec<String>>()
                       .join("\n");
            if list.len() > 0 {
                reason.push_str(" for the following reasons\n");
                reason.push_str(list.as_str());
            }
            ctx.add_rejection_reason(reason);
        }

        Ok(())
    }
}

struct BannedUsernameRejector(SqlPool);

#[async_trait]
impl Validator for BannedUsernameRejector {
    async fn validate(&self, ctx: &mut context::ValidationContext) -> Result<()> {
        let name_bans = ValidationBan::fetch_by_name(ctx.member().guild_id,
                                                     ctx.member().user.name.as_str())
                                      .fetch_all(&self.0)
                                      .await?;
        for ban in name_bans {
            let mut reason = format!("Exact username match with banned user: {}#{}",
                                     ban.name, ban.discriminator);
            if let Some(ban_reason) = ban.reason {
                reason.push_str(format!(" (Ban Reason: {})", ban_reason).as_str());
            }
            ctx.add_rejection_reason(reason);
        }

        if ctx.member().user.avatar.is_none() {
            return Ok(());
        }

        let avatar_bans = ValidationBan::fetch_by_name(ctx.member().guild_id,
                                                       ctx.member().user.avatar.as_ref().unwrap())
                                        .fetch_all(&self.0)
                                        .await?;
        for ban in avatar_bans {
            let mut reason = format!("Exact avatar match with banned user: {}#{}",
                                     ban.name, ban.discriminator);
            if let Some(ban_reason) = ban.reason {
                reason.push_str(format!(" (Ban Reason: {})", ban_reason).as_str());
            }
            ctx.add_rejection_reason(reason);
        }

        Ok(())
    }
}

#[async_trait]
pub trait StringMatchRejector : Sync {
    type Key;
    fn regexes(&self) -> Vec<(Self::Key, Regex)>;
    async fn criteria(&self, ctx: &context::ValidationContext) -> Result<Vec<String>>;
    fn reason(&self, key: &Self::Key, matched: &str) -> String;
}

#[async_trait]
impl<T : StringMatchRejector> Validator for T {
    async fn validate(&self, ctx: &mut context::ValidationContext) -> Result<()> {
        let criteria = self.criteria(&ctx).await?;
        let regexes = self.regexes();
        for check in criteria{
            for (key, regex) in &regexes {
                if let Some(reg_match) = regex.find(check.as_str()) {
                    let reason = self.reason(&key, check.as_str());
                    ctx.add_rejection_reason(reason);
                }
            }
        }

        Ok(())
    }
}

pub struct UsernameMatchRejector {
    sql: SqlPool,
    matches: DashMap<String, Regex>,
    prefix: String
}

impl UsernameMatchRejector {
    pub fn new(
        sql: SqlPool,
        prefix: impl Into<String>,
        matches: Vec<String>
    ) -> Result<Self> {
        Ok(Self {
            sql,
            matches: Self::compile(matches)?,
            prefix: prefix.into()
        })
    }

    fn compile(base: Vec<String>) -> Result<DashMap<String, Regex>> {
        let matches: DashMap<String, Regex> = DashMap::new();
        for input in base {
            matches.insert(input.clone(), Regex::new(Self::generalize_filter(input).as_str())?);
        }
        Ok(matches)
    }

    fn generalize_filter(base: String) -> String {
        base.chars()
            .flat_map(|ch| {
                if ch.is_alphanumeric() {
                    vec![ch, '+']
                } else {
                    vec![ch]
                }
            })
            .collect()
    }
}

#[async_trait]
impl StringMatchRejector for UsernameMatchRejector {
    type Key = String;

    fn regexes(&self) -> Vec<(Self::Key, Regex)> {
        self.matches.iter().map(|kv| (kv.key().clone(), kv.value().clone())).collect()
    }

    async fn criteria(&self, ctx: &context::ValidationContext) -> Result<Vec<String>> {
        Ok(Username::fetch(ctx.member().user.id, Some(20))
                     .fetch_all(&self.sql)
                     .await?
                     .into_iter()
                     .map(|un| un.name)
                     .collect())
    }

    fn reason(&self, key: &Self::Key, matched: &str) -> String {
        format!("{} (Matches: {}): {}", self.prefix, key, matched)
    }
}

pub(super) fn new_account(lookback: Duration) -> BoxedValidator {
    let human_lookback = humantime::format_duration(lookback.to_std().unwrap());
    GenericValidator::new_rejector(
        format!("Account created less than {} ago.", human_lookback),
        move |ctx| Ok(ctx.member().created_at() - Utc::now() < lookback))
}

pub(super) fn no_avatar() -> BoxedValidator {
    GenericValidator::new_rejector(
        "User has no avatar.",
        move |ctx| Ok(ctx.member().user.avatar.is_none()))
}

pub(super) fn banned_user(sql: SqlPool, cache: InMemoryCache, min_guild_size: u64) ->
    BoxedValidator {
    Box::new(BannedUserRejector { sql, cache, min_guild_size })
}

pub(super) fn banned_username(sql: SqlPool) -> BoxedValidator {
    Box::new(BannedUsernameRejector(sql))
}

pub(super) fn deleted_user(sql: SqlPool) -> BoxedValidator {
    Box::new(DeletedUserRejector(sql))
}
